File: src/store/artifacts.ts
import { mkdir, open, rename, stat, writeFile } from "node:fs/promises";
import path from "node:path";
import type { ArtifactRecord } from "../types.js";
import { sha256Bytes } from "../util/hash.js";

export class ArtifactStore {
  constructor(readonly root: string) {}

  async write(content: Uint8Array | string, mediaType: string, producer?: string): Promise<ArtifactRecord> {
    const bytes = typeof content === "string" ? Buffer.from(content, "utf8") : Buffer.from(content);
    const sha256 = sha256Bytes(bytes);
    const directory = path.join(this.root, "sha256");
    await mkdir(directory, { recursive: true });
    const finalPath = path.join(directory, sha256);
    const tempPath = `${finalPath}.${process.pid}.${Date.now()}.tmp`;
    const handle = await open(tempPath, "w");
    try {
      await handle.writeFile(bytes);
      await handle.sync();
    } finally {
      await handle.close();
    }
    await rename(tempPath, finalPath).catch(async (error: NodeJS.ErrnoException) => {
      if (error.code === "EEXIST") {
        await writeFile(tempPath, "");
        return;
      }
      throw error;
    });
    const size = (await stat(finalPath)).size;
    const record: ArtifactRecord = {
      id: sha256,
      sha256,
      mediaType,
      size,
      path: finalPath,
      redactionStatus: "none"
    };
    if (producer !== undefined) record.producer = producer;
    return record;
  }
}

File: src/artifacts.rs
use crate::canonical::sha256_bytes;
use crate::types::ArtifactRecord;
use anyhow::Result;
use std::fs;
use std::io::Write;
use std::path::{Path, PathBuf};

#[derive(Debug, Clone)]
pub struct ArtifactStore {
    root: PathBuf,
}

impl ArtifactStore {
    pub fn new(root: impl Into<PathBuf>) -> Self {
        Self { root: root.into() }
    }

    pub fn write(
        &self,
        content: impl AsRef<[u8]>,
        media_type: &str,
        producer: Option<&str>,
    ) -> Result<ArtifactRecord> {
        let bytes = content.as_ref();
        let sha256 = sha256_bytes(bytes);
        let directory = self.root.join("sha256");
        fs::create_dir_all(&directory)?;
        let final_path = directory.join(&sha256);
        if !final_path.exists() {
            let temp_path = directory.join(format!("{sha256}.{}.tmp", std::process::id()));
            let mut file = fs::File::create(&temp_path)?;
            file.write_all(bytes)?;
            file.sync_all()?;
            fs::rename(&temp_path, &final_path)?;
        }
        let size = fs::metadata(&final_path)?.len();
        Ok(ArtifactRecord {
            id: sha256.clone(),
            sha256,
            media_type: media_type.to_string(),
            size,
            path: final_path.to_string_lossy().into_owned(),
            producer: producer.map(str::to_string),
            redaction_status: "none".to_string(),
        })
    }

    pub fn root(&self) -> &Path {
        &self.root
    }
}
